File: packages/java/src/main/java/io/axene/mailer/Emails.java
package io.axene.mailer;

import com.fasterxml.jackson.databind.JavaType;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Send, look up, search, schedule, and inspect messages. Accessed as {@code client.emails()}. */
public final class Emails {

    private final ApiTransport transport;

    Emails(ApiTransport transport) {
        this.transport = transport;
    }

    /**
     * Send a single email.
     *
     * @param email the message to send.
     * @return the queued message id and its initial status.
     */
    public SendEmailResult send(SendEmail email) {
        return transport.request("POST", "/v1/emails/", email.toWire(), transport.type(SendEmailResult.class));
    }

    /**
     * Send up to your plan's batch limit in one call. The API accepts a bare
     * array of messages and returns a per-message result set.
     *
     * @param emails the messages to send.
     * @return the batch totals and one result per message.
     */
    public BatchResult sendBatch(List<SendEmail> emails) {
        List<Map<String, Object>> wire = new ArrayList<>();
        for (SendEmail e : emails) {
            wire.add(e.toWire());
        }
        return transport.request("POST", "/v1/emails/batch", wire, transport.type(BatchResult.class));
    }

    /**
     * Dry-run a send: check whether {@code message} would be accepted (sender
     * registered, domain verified, plan limits, account not restricted) without
     * actually sending it.
     *
     * @param message the message to validate.
     * @return the validation outcome and current quota usage.
     */
    public ValidationResult validate(SendEmail message) {
        return transport.request("POST", "/v1/emails/validate", message.toWire(), transport.type(ValidationResult.class));
    }

    /**
     * List recent emails, newest first.
     *
     * @param status optional status filter, or null.
     * @param page   zero-based page index.
     * @param limit  page size (1-100).
     * @return a page of email records.
     */
    public List<EmailRecord> list(String status, int page, int limit) {
        String qs = Query.of().add("status", status).add("page", page).add("limit", limit).build();
        return transport.request("GET", "/v1/emails/" + qs, null, transport.listType(EmailRecord.class));
    }

    /**
     * Fetch a single email with its bodies and events.
     *
     * @param id the email id.
     * @return the full email record.
     */
    public EmailDetail get(String id) {
        return transport.request("GET", "/v1/emails/" + Query.enc(id), null, transport.type(EmailDetail.class));
    }

    /**
     * List delivery, open, click, and bounce events for an email.
     *
     * @param id the email id.
     * @return the events, oldest first.
     */
    public List<EmailEvent> events(String id) {
        return transport.request("GET", "/v1/emails/" + Query.enc(id) + "/events", null, transport.listType(EmailEvent.class));
    }

    /**
     * Re-send a bounced, rejected, or failed email as a new message.
     *
     * @param id the original email id.
     * @return the new queued message.
     */
    public SendEmailResult retry(String id) {
        return transport.request("POST", "/v1/emails/" + Query.enc(id) + "/retry", null, transport.type(SendEmailResult.class));
    }

    /**
     * Search emails. {@code q} supports inline tokens ({@code to:}, {@code from:},
     * {@code status:}, {@code domain:}, {@code tag:}); leftover words match as free text.
     *
     * @param q      query string, or null.
     * @param status optional status filter, or null.
     * @param tag    optional tag filter, or null.
     * @param page   zero-based page index.
     * @param limit  page size.
     * @return the matching search hits.
     */
    public List<EmailSearchHit> search(String q, String status, String tag, int page, int limit) {
        String qs = Query.of().add("q", q).add("status", status).add("tag", tag)
                .add("page", page).add("limit", limit).build();
        return transport.request("GET", "/v1/emails/search" + qs, null, transport.listType(EmailSearchHit.class));
    }

    /**
     * List emails scheduled for future delivery, soonest first.
     *
     * @return the scheduled emails.
     */
    public List<ScheduledEmail> listScheduled() {
        return transport.request("GET", "/v1/emails/scheduled", null, transport.listType(ScheduledEmail.class));
    }

    /**
     * Cancel a scheduled email.
     *
     * @param id the scheduled email id.
     * @return the id and its new {@code cancelled} status.
     */
    public StatusResult cancelScheduled(String id) {
        return transport.request("DELETE", "/v1/emails/scheduled/" + Query.enc(id), null, transport.type(StatusResult.class));
    }

    /**
     * Send a scheduled email immediately instead of waiting.
     *
     * @param id the scheduled email id.
     * @return the id and its new {@code queued} status.
     */
    public StatusResult sendScheduledNow(String id) {
        return transport.request("POST", "/v1/emails/scheduled/" + Query.enc(id) + "/send-now", null, transport.type(StatusResult.class));
    }

    /**
     * Poll for emails whose status changed at or after {@code since}. Capped at
     * 50 rows; use for live status updates.
     *
     * @param since required ISO 8601 instant.
     * @return the changed email records.
     */
    public List<EmailRecord> updates(Instant since) {
        String qs = Query.of().add("since", since.toString()).build();
        return transport.request("GET", "/v1/emails/updates" + qs, null, transport.listType(EmailRecord.class));
    }

    /**
     * Get the caller's saved searches.
     *
     * @return the saved searches as opaque maps.
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> getSavedSearches() {
        JavaType type = transport.mapper().getTypeFactory().constructMapType(
                java.util.LinkedHashMap.class, String.class, Object.class);
        Map<String, Object> r = transport.request("GET", "/v1/emails/saved-searches", null, type);
        Object searches = r.get("searches");
        return searches == null ? new ArrayList<>() : (List<Map<String, Object>>) searches;
    }

    /**
     * Replace the caller's saved searches (max 50).
     *
     * @param searches the saved searches to store.
     * @return the normalized saved searches.
     */
    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> setSavedSearches(List<Map<String, Object>> searches) {
        Map<String, Object> body = Wire.map();
        body.put("searches", searches);
        JavaType type = transport.mapper().getTypeFactory().constructMapType(
                java.util.LinkedHashMap.class, String.class, Object.class);
        Map<String, Object> r = transport.request("PUT", "/v1/emails/saved-searches", body, type);
        Object out = r.get("searches");
        return out == null ? new ArrayList<>() : (List<Map<String, Object>>) out;
    }
}

File: packages/go/emails.go
package axene

import (
	"context"
	"net/url"
	"strconv"
)

// Emails is the emails resource, reached as client.Emails.
type Emails struct {
	http *transport
}

// Send sends a single email.
func (e *Emails) Send(ctx context.Context, msg SendEmail) (*SendResult, error) {
	var out SendResult
	if err := e.http.doRequest(ctx, "POST", "/v1/emails/", msg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SendBatch sends up to the plan's batch limit in one call. The API accepts a
// bare array of messages and returns a per-message result set.
func (e *Emails) SendBatch(ctx context.Context, msgs []SendEmail) (*BatchResult, error) {
	var out BatchResult
	if err := e.http.doRequest(ctx, "POST", "/v1/emails/batch", msgs, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Validate dry-runs a send: it checks whether msg would be accepted (sender
// registered, domain verified, plan limits) without actually sending it.
func (e *Emails) Validate(ctx context.Context, msg SendEmail) (*ValidationResult, error) {
	var out ValidationResult
	if err := e.http.doRequest(ctx, "POST", "/v1/emails/validate", msg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List returns recent emails, newest first.
func (e *Emails) List(ctx context.Context, params ListEmailsParams) ([]Email, error) {
	q := url.Values{}
	if params.Status != "" {
		q.Set("status", params.Status)
	}
	q.Set("page", strconv.Itoa(params.Page))
	if params.Limit > 0 {
		q.Set("limit", strconv.Itoa(params.Limit))
	}
	var out []Email
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/?"+q.Encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Get fetches a single email with its bodies and events.
func (e *Emails) Get(ctx context.Context, id string) (*EmailDetail, error) {
	var out EmailDetail
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/"+url.PathEscape(id), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Events lists delivery, open, click, and bounce events for an email.
func (e *Emails) Events(ctx context.Context, id string) ([]EmailEvent, error) {
	var out []EmailEvent
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/"+url.PathEscape(id)+"/events", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Retry re-sends a bounced, rejected, or failed email as a new message.
func (e *Emails) Retry(ctx context.Context, id string) (*SendResult, error) {
	var out SendResult
	if err := e.http.doRequest(ctx, "POST", "/v1/emails/"+url.PathEscape(id)+"/retry", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Search searches emails. q supports inline tokens (to:, from:, status:,
// domain:, tag:); leftover words are matched as free text.
func (e *Emails) Search(ctx context.Context, params SearchEmailsParams) ([]EmailSearchHit, error) {
	q := url.Values{}
	q.Set("q", params.Q)
	if params.Status != "" {
		q.Set("status", params.Status)
	}
	if params.Tag != "" {
		q.Set("tag", params.Tag)
	}
	q.Set("page", strconv.Itoa(params.Page))
	if params.Limit > 0 {
		q.Set("limit", strconv.Itoa(params.Limit))
	}
	var out []EmailSearchHit
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/search?"+q.Encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// ListScheduled lists emails scheduled for future delivery, soonest first.
func (e *Emails) ListScheduled(ctx context.Context) ([]ScheduledEmail, error) {
	var out []ScheduledEmail
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/scheduled", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// CancelScheduled cancels a scheduled email.
func (e *Emails) CancelScheduled(ctx context.Context, id string) (*IDStatus, error) {
	var out IDStatus
	if err := e.http.doRequest(ctx, "DELETE", "/v1/emails/scheduled/"+url.PathEscape(id), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SendScheduledNow sends a scheduled email immediately instead of waiting.
func (e *Emails) SendScheduledNow(ctx context.Context, id string) (*IDStatus, error) {
	var out IDStatus
	if err := e.http.doRequest(ctx, "POST", "/v1/emails/scheduled/"+url.PathEscape(id)+"/send-now", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Updates polls for emails whose status changed at or after since (ISO 8601).
// The result is capped at 50 rows. since is required.
func (e *Emails) Updates(ctx context.Context, since string) ([]Email, error) {
	q := url.Values{}
	q.Set("since", since)
	var out []Email
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/updates?"+q.Encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// savedSearchEnvelope is the wire shape for the saved-searches endpoints.
type savedSearchEnvelope struct {
	Searches []SavedSearch `json:"searches"`
}

// GetSavedSearches returns the caller's saved email searches.
func (e *Emails) GetSavedSearches(ctx context.Context) ([]SavedSearch, error) {
	var out savedSearchEnvelope
	if err := e.http.doRequest(ctx, "GET", "/v1/emails/saved-searches", nil, &out); err != nil {
		return nil, err
	}
	return out.Searches, nil
}

// SetSavedSearches replaces the caller's saved email searches (max 50).
func (e *Emails) SetSavedSearches(ctx context.Context, searches []SavedSearch) ([]SavedSearch, error) {
	var out savedSearchEnvelope
	if err := e.http.doRequest(ctx, "PUT", "/v1/emails/saved-searches", savedSearchEnvelope{Searches: searches}, &out); err != nil {
		return nil, err
	}
	return out.Searches, nil
}
